Initial ClojureScript support

alessandrod opened the pull request:
This PR adds basic ClojureScript.  To test it you can do something like:

```
  $ lein repl
  nREPL server started on port 57872 on host 127.0.0.1 - nrepl://127.0.0.1:57872
  REPL-y 0.3.7, nREPL 0.2.10
  Clojure 1.8.0
  Java HotSpot(TM) 64-Bit Server VM 1.8.0_77-b03
      Docs: (doc function-name-here)
            (find-doc "part-of-name-here")
    Source: (source function-name-here)
   Javadoc: (javadoc java-object-or-class-here)
      Exit: Control+D or (exit) or (quit)
   Results: Stored in vars *1, *2, *3, an exception in *e

  user=> (require 'cljs.repl.node)
  nil
  user=> (cemerick.piggieback/cljs-repl (cljs.repl.node/repl-env))
  ClojureScript Node.js REPL server listening on 52579
  To quit, type: :cljs/quit
```

Then connect to the REPL from vscode, open a `.cljs` file and eval it.

Mind this requires `cider-nrepl 0.15.0-SNAPSHOT` which includes a change I did to enable stacktrace support for clojurescript.

alessandrod committed:
Add optional session parameter to nreplClient::clone()

alessandrod committed:
Add session parameter to nreplClient::{evaluate,evaluateFile}

alessandrod committed:
Initial ClojureScript support.

This change adds basic ClojureScript support. When connecting to the REPL we
look for a CLJS session and if we find one we store it and use it to eval
ClojureScript code.

alessandrod committed:
Add session parameter to nreplClient::close()

avli commented:
Hi @alessandrod 

First of all, thank you for the pull request – ClojureScript support is something many people want this extension to support. Unfortunately, I don't have a lot of experience with ClojureScript and that makes it hard for me to do it by myself.

I tried to test the changes but can't connect to the nREPL instance. Here is what I did:

1. Created a new ClojureScript application with `lein new mies hello-world`.
2. `cd hello-world` and `lein repl`.
3. Did exactly what you wrote inside the REPL.

Now when I try to eval a `cljs` file I get the "No ClojureScript session found in the current nREPL." message. Am I missing something?

Thanks again!

alessandrod commented:
Thanks for looking at this so quickly!

That message is shown when a cljs session isn't found in the nREPL. That can happen if for whatever reason starting the cljs REPL failed, or if you connected before starting the cljs REPL. I have a WIP fix for the latter.

When you tried yesterday I had forgot to specify in the README that you need to depend on Piggieback for cljs to work. I have now fixed that and updated the PR, see [README.md](https://github.com/alessandrod/clojureVSCode/blob/8d110192465769ee0fb3790a830001d5be19b6bc/README.md)

I just followed the updated walkthrough using the mies template and it seems to work. Can you please try again?

avli commented:
Now I'm able to connect to nREPL but got the following message when running the `Eval` command:

```
lojure.lang.Compiler$CompilerException java.io.FileNotFoundException: Could not locate clojure/repl/browser__init.class or clojure/repl/browser.clj on classpath.
 at /private/tmp/hello-world/src/hello_world/core.cljs:0:0
    clojure.lang.Compiler.load (Compiler.java:7391)
    clojure.lang.Compiler.eval (Compiler.java:6927)
    clojure.lang.Compiler.eval (Compiler.java:6890)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate$fn__648.invoke (interruptible_eval.clj:87)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate.invokeStatic (interruptible_eval.clj:85)
    clojure.tools.nrepl.middleware.interruptible_eval$evaluate.invoke (interruptible_eval.clj:55)
    clojure.tools.nrepl.middleware.interruptible_eval$interruptible_eval$fn__693$fn__696.invoke (interruptible_eval.clj:222)
    clojure.tools.nrepl.middleware.interruptible_eval$run_next$fn__688.invoke (interruptible_eval.clj:190)
```

Sorry, maybe I'm doing something wrong? :-(

It's also strange that VSCode doesn't complete things like `enable-console-print!` which, I think, should be available in the ClojureScript `user` namespace. Does autocompletion of ClojureScript works for you?

fasfsfgs commented:
This PR broke stacktrace links. I believe it's because we are now calling `nrepl.eval` instead of `nrepl.evalFile`, which I completely agree method naming wise.
But without using evalFile, the stacktrace doesn't provide good links to files which have compilation problems.

To test, just eval a piece of code that breaks and check the stacktrace file location.
It now gives:
```
clojure.lang.Compiler$CompilerException java.lang.RuntimeException: Unable to resolve symbol: a in this context
 at /private/var/folders/50/5tmv5mjs6gl_n166rxfmb4s40000gn/T/form-init27530814026577044.clj:8:0
    clojure.lang.Compiler.analyze (Compiler.java:6688)
    clojure.lang.Compiler.analyze (Compiler.java:6625)
```

`/private/var/folders/50/5tmv5mjs6gl_n166rxfmb4s40000gn/T/form-init27530814026577044.clj:8:0` is probably a random created file from repl and doesn't serve as a good link resource for the user.

What are the reasons to use eval instead of evalFile? Does it have something to do with cljs?

alessandrod commented:
> What are the reasons to use eval instead of evalFile? Does it have something to do with cljs?

Oh thanks for catching this. I've just fixed it and updated the PR. The relevant change was:

```
    let response;
    if (isSelection && isCljs) {
        // Piggieback's evalFile() ignores the text sent as part of the request
        // and just loads the whole file content from disk. So we use eval()
        // here, which as a drawback will give us a random temporary filename in
        // the stacktrace should an exception occur.
        response = nrepl.eval(ns, text, session);
    } else {
        response = nrepl.evalFile(text, filename, session);
    }`
```

alessandrod commented:
> Now I'm able to connect to nREPL but got the following message when running the Eval command:
> 
> ```
> Clojure.lang.Compiler$CompilerException java.io.FileNotFoundException: Could not locate clojure/repl/browser__init.class or clojure/repl/browser.clj on classpath.
>  at /private/tmp/hello-world/src/hello_world/core.cljs:0:0
>     clojure.lang.Compiler.load (Compiler.java:7391)
>     ...
> ```

Hm that's a bit odd: it's evaluating the ClojureScript file as regular Clojure which is what would happen before my changes.

Are you sure you don't have another version of the extension that is being activated? Can you confirm that you see one of `no ClojureScript session found
` or `found ClojureScript session 92a4dedb-3881-445d-bd93-2c274e21c7f2` in the debug console after connecting to the REPL?

avli commented:
@alessandrod You're right, I run an old version.

Unfortunately, even after ensuring that I run the latest version from your branch I get a compilation error. I believe it's because I'm doing something wrong with the ClojureScript project setup. Could you supply me with a set of instructions about how to set up a project and test it?

I can't say anything about what the compilation error is because that's what I see in the output pane:

<img width="181" alt="screen shot 2017-06-23 at 19 57 37" src="https://user-images.githubusercontent.com/831847/27490147-ffae72b2-585e-11e7-8a4a-0cf171cac44a.png">

@fasfsfgs the picture above looks familiar, do you have any ideas?

avli commented:
All right, I managed to set up a ClojureScript project and successfully evaluate code from inside the editor.

What doesn't work is autocompletion of ClojureScript stuff. @alessandrod is it by the design or it expected to work?

alessandrod commented:
> All right, I managed to set up a ClojureScript project and successfully evaluate code from inside the editor.

Great :)
 
> What doesn't work is autocompletion of ClojureScript stuff. @alessandrod is it by the design or it expected to work?

Yes both the signature and the definition providers should work. I have noticed however that sometimes they don't work for me and it happens with both Clojure and ClojureScript. I did some debugging a couple of days ago and when they don't work, the REPL doesn't get "info" requests at all, so it _seems_ like something may be getting stuck inside the vscode main process.

avli commented:
@alessandrod I continue to test your pull request and found a few other examples of unexpected behavior:

1. After the first successful evaluation of code, you can add incorrect code to the file and the extension will report that compilation was successful.
2. As you mentioned, autocompletion works neither for ClojureScript, nor for Clojure.

One more thing. When I run something like `(println "Hello World)`, I see the output it in nREPL but not in the "Output" pane. I'm not sure it's an issue because I have no idea how to do it right for ClojureScript... If you have any ideas, please share them.

fasfsfgs commented:
> @fasfsfgs the picture above looks familiar, do you have any ideas?

@avli That happened before when we tested #30 without [doing proper configuration](https://github.com/avli/clojureVSCode#before-you-start) first. That's what created #31. I guess it's the same thing. It was probably missing some configuration.
@alessandrod added a couple steps on [README](https://github.com/alessandrod/clojureVSCode/blob/master/README.md#before-you-start) to work with cljs.

> As you mentioned, autocompletion works neither for ClojureScript, nor for Clojure.

I just tested production and it's working here. Is it broken in production as well or just this PR?

I'm starting to test this PR now as well.

@alessandrod Can you format your code as well please? We are just using built-in typescript formatting from vscode (shift+alt+f).

avli commented:
@fasfsfgs The production version works fine. The version from this PR work fine too but for Clojure-only projects. As soon as you add some ClojureScript autocompletion stops working both for `.clj` and `.cljs` files.

fasfsfgs commented:
I think I found a couple bugs related to namespace:

- Eval `clj` ignores namespace when using it with text selection. That can cause some issues when using multiple files because everything will use just a single namespace and `require` breaks; 
To reproduce, use 2 namespaces and 1 of them requires the other. Then eval some function definitions from both and try to eval some code that uses the require alias.

- If I eval `(defn foo [bar] (+ bar 1))` in a cljs file without compiling the whole file first, it gives me the compilation error below (with some unwanted NaN at the end of the stacktrace):
```
clojure.lang.ExceptionInfo repl:1
(function (){try{return cljs.core.pr_str.call(null,(function (){var ret__6877__auto__ = .foo = cljs.core.with_meta((function $foo(bar){
                                                                                        ^

SyntaxError: Unexpected token .
    at createScript (vm.js:74:10)
    at Object.runInThisContext (vm.js:116:10)
    at Domain.<anonymous> ([stdin]:41:34)
    at Domain.run (domain.js:242:14)
    at Socket.<anonymous> ([stdin]:40:25)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:235:13)
    at Socket.Readable.push (_stream_readable.js:197:10)
 at undefined:NaN:NaN
    clojure.tools.nrepl.middleware.interruptible_eval$run_next$fn__701.invoke (interruptible_eval.clj:190)
```

alessandrod committed:
Add optional session parameter to nreplClient::info

alessandrod committed:
nreplClient::send filter undefined fields

Optional request message fields are represented as undefined. Strip them out
before encoding and writing messages on the wire.

Update calling code accordingly.

alessandrod commented:
Sorry it took a while to rebase this, so many things changed in master shortly after I started the PR.

Anyway I've fixed the two namespacing issues reported above and tested that hover, definition and signature work for both clj and cljs.

realgenekim commented:
PS: This is super-exciting — happy to test this out when it's available.  Would love to have Clojure/ClojureScript dev environment inside of Visual Studio Code and reduce reliance on Atom.   Thank you!

avli commented:
Hi, @alessandrod 

Thank you for the update. I've tried the changes you sent – many bugs from the previous version have gone – but I some things still doesn't work for me, e.g. code completion. I suspect, my ClojureScript project setup may be missing something and that's what causes the issues.

I've uploaded the ClojrueScript project I used to [this repository](https://github.com/avli/clojurescript-example-project). I run an nREPL instance in the terminal and then execute the following commands:

```
user=> (require 'cljs.repl.node)
user=> (cemerick.piggieback/cljs-repl (cljs.repl.node/repl-env))
```

Everything looks fine. After connecting to the REPL from VSCode I can execute code but code completion works neither for ClojureScript, nor for Clojure.

Could you try my project and confirm it works for you or point out what it misses?

alessandrod commented:
> Could you try my project and confirm it works for you or point out what it misses?

The problem is that it depends on cider-nrepl, but then specifies `:repl-options {:nrepl-middleware [cemerick.piggieback/wrap-cljs-repl]}`. That tells cider-nrepl to only load piggieback and not the other middlewares needed to implement completion etc.

That can be fixed by using the cider-nrepl plugin, which loads the default middlewares *plus* the ones specified in the project map.

Created a PR for it https://github.com/avli/clojurescript-example-project/pull/1

avli commented:
@alessandrod thank you for the explanation. Now I'm able evaluate `.cljs` files and autocompletion works at least for the stuff from the `core.cljs` module. I'm wondering if autocompletion works for you for things like `js/...`?

alessandrod commented:
@avli it does not. Looking at https://github.com/clojure-emacs/cljs-tooling/blob/master/src/cljs_tooling/complete.clj I don't think it's supposed to work.

avli commented:
In this case, we are all set up to release the first version with ClojureScript support!

alessandrod commented:
Thanks everyone for testing this and @avli for merging it!

avli commented:
@alessandrod Thank you! Any further improvements are more than welcome!
